From 5756c09446b47674a050df6c112b529fead5329b Mon Sep 17 00:00:00 2001
From: Jeff Forcier <jeff@bitprophet.org>
Date: Fri, 17 Jan 2020 15:29:53 -0500
Subject: [PATCH] Drop support for Sphinx <1.8

---
 .travis.yml        |  3 +--
 README.rst         |  2 +-
 docs/changelog.rst | 10 ++++++++
 releases/util.py   | 60 ++++++++++------------------------------------
 setup.py           |  2 +-
 5 files changed, 25 insertions(+), 52 deletions(-)

#diff --git a/.travis.yml b/.travis.yml
#index 9e4bd84..3a8f964 100644
#--- a/.travis.yml
#+++ b/.travis.yml
#@@ -8,8 +8,7 @@ python:
#   - "pypy"
#   #- "pypy3" # Looks like Sphinx (as of 1.4.1) is not pypy3 compat
# env:
#-  - SPHINX=">=1.3,<1.4"
#-  - SPHINX=">=1.7,<1.8"
#+  - SPHINX=">=1.8,<2.0"
#   - SPHINX=">=2.3,<2.4"
# jobs:
#   exclude:
diff --git a/README.rst b/README.rst
index bdaac6b..d4cbb3e 100644
--- a/README.rst
+++ b/README.rst
@@ -5,7 +5,7 @@ What is Releases?
 =================
 
 Releases is a Python (2.7, 3.4+) compatible `Sphinx <http://sphinx-doc.org>`_
-(1.3+) extension designed to help you keep a source control friendly, merge
+(1.8+) extension designed to help you keep a source control friendly, merge
 friendly changelog file & turn it into useful, human readable HTML output.
 
 Specifically:
diff --git a/docs/changelog.rst b/docs/changelog.rst
index 20d813e..72c7648 100644
--- a/docs/changelog.rst
+++ b/docs/changelog.rst
@@ -2,6 +2,16 @@
 Changelog
 =========
 
+- :support:`-` Dropped support for Sphinx <1.8, which is now pretty rare in the
+  wild. This makes it easier to support Sphinx 1.8+ and lets us drop an
+  ever-growing amount of compatibility code for Sphinx 1.3-1.7.
+
+  .. warning::
+    This change is technically backwards incompatible, but our own API and
+    behavior is not changing, and Sphinx itself did not make breaking changes
+    (that we're aware of) in 1.8, so as long as you upgrade your Sphinx along
+    with your Releases, you should be okay.
+
 - :release:`1.6.3 <2020-01-10>`
 - :support:`87 backported` (via :issue:`88`) Our upper Sphinx version limit was
   mostly defensive and at this point is just too old to even build on
diff --git a/releases/util.py b/releases/util.py
index 12e3365..5a069b3 100644
--- a/releases/util.py
+++ b/releases/util.py
@@ -11,25 +11,10 @@
 from docutils.io import NullOutput
 from docutils.nodes import bullet_list
 from sphinx.application import Sphinx # not exposed at top level
-try:
-    from sphinx.io import (
-        SphinxStandaloneReader, SphinxFileInput, SphinxDummyWriter,
-    )
-except ImportError:
-    # NOTE: backwards compat with Sphinx 1.3
-    from sphinx.environment import (
-        SphinxStandaloneReader, SphinxFileInput, SphinxDummyWriter,
-    )
-# sphinx_domains is only in Sphinx 1.5+, but is presumably necessary from then
-# onwards.
-try:
-    from sphinx.util.docutils import sphinx_domains
-except ImportError:
-    # Just dummy it up.
-    from contextlib import contextmanager
-    @contextmanager
-    def sphinx_domains(env):
-        yield
+from sphinx.io import (
+    SphinxStandaloneReader, SphinxFileInput, SphinxDummyWriter,
+)
+from sphinx.util.docutils import sphinx_domains
 
 from . import construct_releases, setup
 
@@ -134,35 +119,22 @@ def get_doctree(path, **kwargs):
     # Create & init a BuildEnvironment. Mm, tasty side effects.
     app._init_env(freshenv=True)
     env = app.env
-    # More arity/API changes: Sphinx 1.3/1.4-ish require one to pass in the app
-    # obj in BuildEnvironment.update(); modern Sphinx performs that inside
-    # Application._init_env() (which we just called above) and so that kwarg is
-    # removed from update(). EAFP.
-    kwargs = dict(
+    env.update(
         config=app.config,
         srcdir=root,
         doctreedir=app.doctreedir,
-        app=app,
     )
-    try:
-        env.update(**kwargs)
-    except TypeError:
-        # Assume newer Sphinx w/o an app= kwarg
-        del kwargs['app']
-        env.update(**kwargs)
+    # Update "temp" data (must be done here as it's wiped on update())
+    env.temp_data['docname'] = docname
     # Code taken from sphinx.environment.read_doc; easier to manually call
     # it with a working Environment object, instead of doing more random crap
     # to trick the higher up build system into thinking our single changelog
     # document was "updated".
-    env.temp_data['docname'] = docname
     env.app = app
-    # NOTE: SphinxStandaloneReader API changed in 1.4 :(
     reader_kwargs = {
         'app': app,
         'parsers': env.config.source_parsers,
     }
-    if sphinx.version_info[:2] < (1, 4):
-        del reader_kwargs['app']
     # This monkeypatches (!!!) docutils to 'inject' all registered Sphinx
     # domains' roles & so forth. Without this, rendering the doctree lacks
     # almost all Sphinx magic, including things like :ref: and :doc:!
@@ -254,12 +226,10 @@ def make_app(**kwargs):
     load_extensions = kwargs.pop('load_extensions', False)
     real_conf = None
     try:
-        # Sphinx <1.6ish
-        Sphinx._log = lambda self, message, wfile, nonl=False: None
-        # Sphinx >=1.6ish. Technically still lets Very Bad Things through,
-        # unlike the total muting above, but probably OK.
-        # NOTE: used to just do 'sphinx' but that stopped working, even on
-        # sphinx 1.6.x. Weird. Unsure why hierarchy not functioning.
+        # Turn off most logging, which is rarely useful and usually just gums
+        # up the output of whatever tool is calling us.
+        # NOTE: used to just do 'sphinx' but that stopped working. Unsure why
+        # hierarchy not functioning.
         for name in ('sphinx', 'sphinx.sphinx.application'):
             logging.getLogger(name).setLevel(logging.ERROR)
         # App API seems to work on all versions so far.
@@ -300,13 +270,7 @@ def make_app(**kwargs):
         config['releases_{}'.format(name)] = kwargs[name]
     # Stitch together as the sphinx app init() usually does w/ real conf files
     app.config._raw_config = config
-    # init_values() requires a 'warn' runner on Sphinx 1.3-1.6, so if we seem
-    # to be hitting arity errors, give it a dummy such callable. Hopefully
-    # calling twice doesn't introduce any wacko state issues :(
-    try:
-        app.config.init_values()
-    except TypeError: # boy I wish Python had an ArityError or w/e
-        app.config.init_values(lambda x: x)
+    app.config.init_values()
     # Initialize extensions (the internal call to this happens at init time,
     # which of course had no valid config yet here...)
     if load_extensions:
diff --git a/setup.py b/setup.py
index 10b17eb..6cefb55 100644
--- a/setup.py
+++ b/setup.py
@@ -19,7 +19,7 @@
     packages=['releases'],
     install_requires=[
         'semantic_version<2.7',
-        'sphinx>=1.3',
+        'sphinx>=1.8',
     ],
     classifiers=[
         'Development Status :: 5 - Production/Stable',
From fd620847a179c2587b75c710925bc71158258e01 Mon Sep 17 00:00:00 2001
From: Jeff Forcier <jeff@bitprophet.org>
Date: Fri, 17 Jan 2020 21:10:04 -0500
Subject: [PATCH] sphinx.io.read_doc now does exactly what we needed

No idea when they added it, hopefully after I wrote
all those gross glarly hacks. And good riddance!
---
 releases/util.py | 59 ++++--------------------------------------------
 1 file changed, 4 insertions(+), 55 deletions(-)

diff --git a/releases/util.py b/releases/util.py
index 661bd46..3c3e810 100644
--- a/releases/util.py
+++ b/releases/util.py
@@ -6,15 +6,9 @@
 import os
 from tempfile import mkdtemp
 
-import sphinx
-from docutils.core import Publisher
-from docutils.io import NullOutput
 from docutils.nodes import bullet_list
 from sphinx.application import Sphinx # not exposed at top level
-from sphinx.io import (
-    SphinxStandaloneReader, SphinxFileInput, SphinxDummyWriter,
-)
-from sphinx.util.docutils import sphinx_domains
+from sphinx.io import read_doc
 
 from . import construct_releases, setup
 
@@ -118,54 +112,9 @@ def get_doctree(path, **kwargs):
     # TODO: this only works for top level changelog files (i.e. ones where
     # their dirname is the project/doc root)
     app = make_app(srcdir=root, **kwargs)
-    # Create & init a BuildEnvironment. Mm, tasty side effects.
-    app._init_env(freshenv=True)
-    env = app.env
-    env.update(
-        config=app.config,
-        srcdir=root,
-        doctreedir=app.doctreedir,
-    )
-    # Update "temp" data (must be done here as it's wiped on update())
-    env.temp_data['docname'] = docname
-    # Code taken from sphinx.environment.read_doc; easier to manually call
-    # it with a working Environment object, instead of doing more random crap
-    # to trick the higher up build system into thinking our single changelog
-    # document was "updated".
-    env.app = app
-    reader_kwargs = {
-        'app': app,
-        'parsers': env.config.source_parsers,
-    }
-    # This monkeypatches (!!!) docutils to 'inject' all registered Sphinx
-    # domains' roles & so forth. Without this, rendering the doctree lacks
-    # almost all Sphinx magic, including things like :ref: and :doc:!
-    with sphinx_domains(env):
-        try:
-            reader = SphinxStandaloneReader(**reader_kwargs)
-        except TypeError:
-            # If we import from io, this happens automagically, not in API
-            del reader_kwargs['parsers']
-            reader = SphinxStandaloneReader(**reader_kwargs)
-        pub = Publisher(reader=reader,
-                        writer=SphinxDummyWriter(),
-                        destination_class=NullOutput)
-        pub.set_components(None, 'restructuredtext', None)
-        pub.process_programmatic_settings(None, env.settings, None)
-        # NOTE: docname derived higher up, from our given path
-        src_path = env.doc2path(docname)
-        source = SphinxFileInput(
-            app,
-            env,
-            source=None,
-            source_path=src_path,
-            encoding=env.config.source_encoding,
-        )
-        pub.source = source
-        pub.settings._source = src_path
-        pub.set_destination(None, None)
-        pub.publish()
-        return app, pub.document
+    app.env.temp_data['docname'] = docname
+    doctree = read_doc(app, app.env, path)
+    return app, doctree
 
 
 def load_conf(srcdir):
